pallets/file-system/src/lib.rs

```diff
@@ -0,0 +1,310 @@
+            todo!()
+        }
+
+        /// Used by a BSP to volunteer for storing a file.
+        ///
+        /// The transaction will fail if the XOR between the file ID and the BSP ID is not below the threshold,
+        /// so a BSP is strongly advised to check beforehand. Another reason for failure is
+        /// if the maximum number of BSPs has been reached. A successful assignment as BSP means
+        /// that some of the collateral tokens of that MSP are frozen.
+        #[pallet::call_index(4)]
+        #[pallet::weight(10_000 + T::DbWeight::get().reads_writes(1,1).ref_time())]
+        pub fn bsp_volunteer(
+            origin: OriginFor<T>,
+            location: FileLocation<T>,
+            fingerprint: Fingerprint<T>,
+            bsp_multiaddress: MultiAddress<T>,
+        ) -> DispatchResult {
+            // Check that the extrinsic was signed and get the signer.
+            let who = ensure_signed(origin)?;
+
+            // Perform validations and register Storage Provider as BSP for file.
+            Self::do_bsp_volunteer(who.clone(), location.clone(), fingerprint)?;
+
+            // Emit new BSP volunteer event.
+            Self::deposit_event(Event::AcceptedBspVolunteer {
+                who,
+                location,
+                fingerprint,
+                bsp_multiaddress,
+            });
+
+            Ok(())
+        }
+
+        /// Executed by a BSP to stop storing a file.
+        ///
+        /// A compensation should be provided for the user, to deter this behaviour. A successful execution of this extrinsic automatically generates a storage request for that file with one remaining_bsps_slot left, and if a storage request for that file already exists, the slots left are incremented in one. It also automatically registers a challenge for this file, for the next round of storage proofs, so that the other BSPs and MSP who are storing it would be forced to disclose themselves then.
+        #[pallet::call_index(5)]
+        #[pallet::weight(10_000 + T::DbWeight::get().reads_writes(1,1).ref_time())]
+        pub fn bsp_stop_storing(_origin: OriginFor<T>) -> DispatchResult {
+            todo!()
+        }
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T>
+    where
+        u32: TryFrom<BlockNumberFor<T>>,
+    {
+        fn on_idle(block: BlockNumberFor<T>, remaining_weight: Weight) -> Weight {
+            let db_weight = T::DbWeight::get();
+
+            // Return early if the remaining weight is not enough to perform the operation
+            // TODO: fix this, it's not working as expected (when the remaining weight is 0, it stil goes through)
+            if !remaining_weight
+                .all_gte(db_weight.reads_writes(1, T::MaxExpiredStorageRequests::get().into()))
+            {
+                CurrentExpirationBlock::<T>::put(block.saturating_add(1u32.into()));
+                return Weight::zero();
```

Comment:
The logics of this `on_idle` still need to be polished, but it is ok for this PR's scope.